test/extensions/filters/http/cache/http_cache_implementation_test_common.cc

```diff
@@ -0,0 +1,747 @@
+
+void HttpCacheImplementationTest::SetUp() {
+  request_headers_.setMethod("GET");
+  request_headers_.setHost("example.com");
+  request_headers_.setScheme("https");
+  request_headers_.setCopy(Http::CustomHeaders::get().CacheControl, "max-age=3600");
+
+  EXPECT_CALL(dispatcher_, post(_)).Times(AnyNumber());
+  EXPECT_CALL(dispatcher_, isThreadSafe()).Times(AnyNumber());
+
+  delegate_->setUp(dispatcher_);
+}
+
+void HttpCacheImplementationTest::TearDown() {
+  Assert::resetEnvoyBugCountersForTest();
+
+  delegate_->tearDown();
+}
+
+void HttpCacheImplementationTest::updateHeaders(
+    absl::string_view request_path, const Http::TestResponseHeaderMapImpl& response_headers,
+    const ResponseMetadata& metadata) {
+  LookupContextPtr lookup_context = lookup(request_path);
+  cache().updateHeaders(*lookup_context, response_headers, metadata);
+}
+
+LookupContextPtr HttpCacheImplementationTest::lookup(absl::string_view request_path) {
+  LookupRequest request = makeLookupRequest(request_path);
+  LookupContextPtr context = cache().makeLookupContext(std::move(request), decoder_callbacks_);
+  absl::Notification lookup_result_received;
+  context->getHeaders([this, &lookup_result_received](LookupResult&& result) {
+    lookup_result_ = std::move(result);
+    lookup_result_received.Notify();
+  });
+  EXPECT_TRUE(lookup_result_received.WaitForNotificationWithTimeout(absl::Seconds(3)));
+  return context;
+}
+
+absl::Status HttpCacheImplementationTest::insert(LookupContextPtr lookup,
+                                                 const Http::TestResponseHeaderMapImpl& headers,
+                                                 const absl::string_view body,
+                                                 std::chrono::milliseconds timeout) {
+  return insert(std::move(lookup), headers, body, absl::nullopt, timeout);
+}
+
+absl::Status HttpCacheImplementationTest::insert(
+    LookupContextPtr lookup, const Http::TestResponseHeaderMapImpl& headers,
+    const absl::string_view body, const absl::optional<Http::TestResponseTrailerMapImpl> trailers,
+    std::chrono::milliseconds timeout) {
+  InsertContextPtr inserter = cache().makeInsertContext(std::move(lookup), encoder_callbacks_);
+  const ResponseMetadata metadata{time_source_.systemTime()};
+  bool headers_end_stream = body.empty() && !trailers.has_value();
+  inserter->insertHeaders(headers, metadata, headers_end_stream);
+
+  if (headers_end_stream) {
+    return absl::OkStatus();
+  }
+  if (!body.empty() && !trailers.has_value()) {
+    inserter->insertBody(Buffer::OwnedImpl(body), nullptr,
+                         /*end_stream=*/!trailers.has_value());
+    return absl::OkStatus();
+  }
+  if (body.empty() && trailers.has_value()) {
+    inserter->insertTrailers(trailers.value());
+    return absl::OkStatus();
+  }
+
+  // For responses with body and trailers, wait for insertBody's callback before
+  // calling insertTrailers.
+  std::promise<absl::Status> insert_promise;
```

Comment:
I think this will be the first use of std::promise in Envoy. Not sure if that's a concern or not.

I'm not sure how it would interact in production code with the siloed/async threading model but maybe it's OK for tests?

I think should probably get someone else to weigh in on this also, but for now I asked on Slack.

In the meantime I have a functional issue with your usage below.